File: efa/Efa.php
<?php

class Efa
{

  public function __construct($arguments){
    $this->arguments = $arguments;
  }

  public function run(){

    if ($this->arguments['help'] || count($_SERVER['argv']) == 1) {
       echo $this->arguments->getHelpScreen().PHP_EOL.PHP_EOL;
       die();
    }

    if($this->arguments['start'] && $this->arguments['ziel'])
       $this->route();

    if($this->arguments['start'] && !$this->arguments['ziel'])
       $this->monitor();

    if($this->arguments['meldungen'])
       $this->meldungen();

     if($this->arguments['info'])
        $this->info();

  }

  public function info(){
    if(!is_numeric($this->arguments['info'])){
      $this->getStationList('info','info');
      return false;
    }
    $departures = json_decode(file_get_contents(sprintf('https://www.efa-bw.de/bvb3/XML_DM_REQUEST?laguage=de&typeInfo_dm=stopID&deleteAssignedStops_dm=1&useRealtime=1&mode=direct&excludedMeans=0&excludedMeans=1&excludedMeans=2&limit=10&outputFormat=JSON&coordOutputFormat=WGS84[DD.DDDDD]&nameInfo_dm=%s', urlencode($this->arguments['info']))), false);
    \cli\line("Efa-ID: {$departures->dm->points->point->ref->id}");
    $geo = explode(',', $departures->dm->points->point->ref->coords);
    \cli\line("Koordinaten: {$geo[1]}, {$geo[0]}");
    \cli\line("Google Maps: https://www.google.com/maps/search/?api=1&query={$geo[1]},{$geo[0]}");
    \cli\line("BVB Anfahrtsmonitor: https://dfi.citykanal.com/?point={$departures->dm->points->point->ref->id}");

  }

  public function route(){

    if(!is_numeric($this->arguments['start'])){
      $this->getStationList('route','start');
      return false;
    }
    if(!is_numeric($this->arguments['ziel'])){
      $this->getStationList('route','ziel');
      return false;
    }

    $res   = file_get_contents(sprintf("https://www.efa-bw.de/bvb3/XSLT_TRIP_REQUEST2?itdDate=%s&itdTime=%s&language=de&sessionID=0&outputFormat=JSON&type_origin=stop&name_origin=%s&type_destination=stop&name_destination=%s", date('Ymd') , date('Hi'), $this->arguments['start'], $this->arguments['ziel']));
    $data  = json_decode(utf8_encode($res) ,false);

    \cli\line("Verfügbare Routen von %s nach %s".PHP_EOL, $data->origin->points->point->name, $data->destination->points->point->name);
    $trips = [];
    $i = 0;
    foreach($data->trips as $trip){
      $trips["t-".$i++] = sprintf("Ab %s um %s mit %sx Umsteigen. Dauer %sh",
                              $trip->legs[0]->points[0]->name,
                              $trip->legs[0]->points[0]->dateTime->time,
                              $trip->interchange,
                              $trip->duration
                            );
    }
    $trips['quit'] = 'Abbrechen';
    while (true) {
      $choice = \cli\menu($trips, null, 'Bitte wähle eine Fahrt');
      \cli\line();
      if ($choice == 'quit') {
        break;
      } else {
         $tabledata = [];
         foreach($data->trips[str_replace('t-','',$choice)]->legs as $legs){
           $tabledata[] = [
             (empty($legs->mode->name)) ? "Fussweg" : $legs->mode->name,
             $legs->points[0]->name,
             $legs->points[0]->dateTime->time."h",
             $legs->points[1]->dateTime->time."h",
             $legs->points[1]->name
           ];
         }
         if($this->arguments['nofancy']){
           foreach($tabledata as $line){
             \cli\line("{:0} {:1} {:2} {:3} {:4}", $line);
           }
         } else {
           $table = new \cli\Table();
           $table->setHeaders(['', 'Einsteigen', 'Abfahrt', 'Ankunft', 'Aussteigen/Umsteigen']);
           $table->setRows($tabledata);
           $table->display();
         }
         break;

      }
    }

  }

  public function monitor(){

    if(is_numeric($this->arguments['start'])){
      $data   = [];
      $headers= ['Min', '', '#', 'Nach'];
      $departures = json_decode(file_get_contents(sprintf('https://www.efa-bw.de/bvb3/XML_DM_REQUEST?laguage=de&typeInfo_dm=stopID&deleteAssignedStops_dm=1&useRealtime=1&mode=direct&excludedMeans=0&excludedMeans=1&excludedMeans=2&limit=10&outputFormat=JSON&nameInfo_dm=%s', urlencode($this->arguments['start']))), false);
      foreach($departures->departureList as $departure){
        $data[] = [
          $this->padd($departure->countdown),
          $departure->servingLine->name,
          $departure->servingLine->number,
          $departure->servingLine->direction
        ];
      }
      if($this->arguments['nofancy']){
        foreach($data as $line){
          \cli\line("{:0} {:1} {:2} {:3}", $line);
        }
      } else {
        $table = new \cli\Table();
        $table->setHeaders($headers);
        $table->setRows($data);
        $table->display();
      }
    } else {
      $this->getStationList('monitor','start');
    }
  }

  public function getStationList($callback, $target){
    $stations = json_decode(file_get_contents(sprintf('https://www.efa-bw.de/bvb3/XSLT_STOPFINDER_REQUEST?language=de&outputFormat=JSON&coordOutputFormat=WGS84[DD.ddddd]&itdLPxx_usage=origin&useLocalityMainStop=true&SpEncId=0&locationServerActive=1&stateless=1&type_sf=any&anyObjFilter_sf=2&anyMaxSizeHitList=50&name_sf=%s', urlencode($this->arguments[$target]))), false);
    if(count($stations->stopFinder->points) == 1){
      $this->arguments[$target] = $stations->stopFinder->points->point->stateless;
      call_user_func(array($this, $callback));
    } else {
      $menu = [];
      foreach($stations->stopFinder->points as $point){
        $menu[$point->stateless] = $point->name;
      }
      $menu['quit'] = 'Abbrechen';
      \cli\line("Parameter --%s hat mehrere Haltestellenmöglichkeiten:".PHP_EOL, $target);
      while (true) {
      	$choice = \cli\menu($menu, null, 'Bitte wähle eine Haltestelle');
      	\cli\line();
      	if ($choice == 'quit') {
      		break;
      	} else {
          $this->arguments[$target] = $choice;
          call_user_func(array($this, $callback));
          break;
        }
      }
    }
  }

  public function meldungen(){
    $data = json_decode(file_get_contents("https://www.efa-bw.de/bvb3/XML_ADDINFO_REQUEST?filterProviderCode=Basler%20Verkehrs-Betriebe%20(BVB)&outputFormat=JSON"), false);
    if($data->additionalInformation->travelInformations->travelInformation){

      $menu = [];
      $message = [];

      foreach($data->additionalInformation->travelInformations->travelInformation as $info){

        $text = $info->infoLink->content;
        $text = str_replace('&nbsp;','', $text);
        $text = html_entity_decode($text);
        $text = strip_tags($text, '<br>');
        $text = str_replace(PHP_EOL,'', $text);
        $text = preg_replace('/(\>)\s*(\<)/m', '$1$2', $text);
        $text = preg_replace('/(<br\ ?\/?>)+/', '<br>', $text);
        $text = preg_replace('/\s+/', ' ', $text);
        $text = str_replace('<br>',PHP_EOL, $text);

        $menu[md5($info->infoLink->infoLinkText)] = $info->infoLink->infoLinkText;
        $message[md5($info->infoLink->infoLinkText)] = $text;

      }
      $menu['quit'] = 'Abbrechen';
      while (true) {
      	$choice = \cli\menu($menu, null, 'Bitte wähle eine Meldung');
      	\cli\line();
      	if ($choice == 'quit') {
      		break;
      	} else {
          \cli\line(\cli\Colors::colorize("%g ✓ {$menu[$choice]} %n ", true));
          \cli\out($message[$choice]);
          break;
        }
      }
    } else {
      \cli\line(\cli\Colors::colorize("%g✓ Aktuell sind keine Störungen bekannt.%n", true));
    }

  }

  private function padd($t){
   return ($t < 10) ? " ".$t."'" : $t."'";
  }

}

File: internal/efa/efa.go
// Package efa replicates the App\Helpers\Efa class of the original
// efa-cli: the HTTP requests to the EFA journey planner API with
// Guzzle-compatible query encoding.
package efa

import (
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"

	"github.com/simonjenny/efa-cli/internal/i18n"
	"github.com/simonjenny/efa-cli/internal/jsonx"
	"github.com/simonjenny/efa-cli/internal/prompts"
)

const baseURL = "https://www.efa-bw.de/bvb3/"

// Client mirrors the App\Helpers\Efa class.
type Client struct {
	httpClient *http.Client
	// BaseURL overrides the default API base URL (used by tests).
	BaseURL string
	// UseSpinner controls whether requests show the "Fetching Data..."
	// spinner (mirrors the json config flag: no spinner in JSON mode).
	UseSpinner bool
}

// NewClient creates a client with the default HTTP settings.
func NewClient() *Client {
	return &Client{
		httpClient: &http.Client{Timeout: 30 * time.Second},
		BaseURL:    baseURL,
	}
}

// guzzleQueryEncode percent-encodes a query value the way Guzzle 7.9
// (PSR-7 Uri::filterQuery) does: unreserved characters, reserved
// characters and valid percent sequences stay as-is; everything else is
// percent-encoded with uppercase hex digits.
func guzzleQueryEncode(s string) string {
	var b strings.Builder
	for i := 0; i < len(s); i++ {
		c := s[i]
		if c == '%' {
			if i+2 < len(s) && isHex(s[i+1]) && isHex(s[i+2]) {
				b.WriteByte(c)
			} else {
				b.WriteString("%25")
			}
			continue
		}
		if isQueryAllowed(c) {
			b.WriteByte(c)
			continue
		}
		fmt.Fprintf(&b, "%%%02X", c)
	}
	return b.String()
}

func isQueryAllowed(c byte) bool {
	switch {
	case c >= 'a' && c <= 'z', c >= 'A' && c <= 'Z', c >= '0' && c <= '9':
		return true
	}
	switch c {
	case '-', '_', '.', '~', '!', '$', '&', '\'', '(', ')', '*', '+', ',', ';', '=', '%', ':', '@', '/', '?':
		return true
	}
	return false
}

func isHex(c byte) bool {
	return c >= '0' && c <= '9' || c >= 'a' && c <= 'f' || c >= 'A' && c <= 'F'
}

type queryParam struct {
	key   string
	value string
}

// buildURL constructs the request URL with the exact query parameter
// order used by the PHP application.
func buildURL(base string, endpoint string, params []queryParam) string {
	var b strings.Builder
	b.WriteString(base)
	b.WriteString(endpoint)
	b.WriteByte('?')
	for i, p := range params {
		if i > 0 {
			b.WriteByte('&')
		}
		b.WriteString(p.key)
		b.WriteByte('=')
		b.WriteString(guzzleQueryEncode(p.value))
	}
	return b.String()
}

// get performs the HTTP GET, decoding the JSON response. Invalid or
// empty responses decode to an empty array, matching PHP's
// response->object() ?? [].
func (c *Client) get(endpoint string, params []queryParam) (any, error) {
	url := buildURL(c.BaseURL, endpoint, params)
	if c.UseSpinner {
		return prompts.SpinResult(i18n.T("spinner.fetching"), func() (any, error) {
			return c.doGet(url)
		})
	}
	return c.doGet(url)
}

func (c *Client) doGet(url string) (any, error) {
	req, err := http.NewRequest(http.MethodGet, url, nil)
	if err != nil {
		return []any{}, nil
	}
	req.Header.Set("Accept", "application/json")
	req.Header.Set("User-Agent", "GuzzleHttp/7")
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return []any{}, nil
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return []any{}, nil
	}
	v, err := jsonx.Decode(body)
	if err != nil {
		return []any{}, nil
	}
	return v, nil
}

// Option is a stop search result (GID + name).
type Option struct {
	Key   string
	Label string
}

// Meldungen returns the BVB travel information messages, mirroring
// Efa::meldungen().
func (c *Client) Meldungen() ([]any, error) {
	v, err := c.get("XML_ADDINFO_REQUEST", []queryParam{
		{key: "filterProviderCode", value: "Basler Verkehrs-Betriebe (BVB)"},
		{key: "outputFormat", value: "JSON"},
	})
	if err != nil {
		return nil, err
	}
	return toAnySlice(jsonx.Path(v, "additionalInformation", "travelInformations", "travelInformation")), nil
}

// Haltestelle resolves a single stop, mirroring Efa::haltestelle().
// It returns nil when no stop could be resolved (PHP: false).
func (c *Client) Haltestelle(name string) (*jsonx.Ordered, error) {
	v, err := c.get("XSLT_STOPFINDER_REQUEST", []queryParam{
		{key: "language", value: "de"},
		{key: "outputFormat", value: "JSON"},
		{key: "coordOutputFormat", value: "WGS84[DD.ddddd]"},
		{key: "itdLPxx_usage", value: "origin"},
		{key: "useLocalityMainStop", value: "true"},
		{key: "SpEncId", value: "0"},
		{key: "locationServerActive", value: "1"},
		{key: "stateless", value: "1"},
		{key: "type_sf", value: "any"},
		{key: "anyObjFilter_sf", value: "2"},
		{key: "anyMaxSizeHitList", value: "1"},
		{key: "name_sf", value: name},
	})
	if err != nil {
		return nil, err
	}
	point := jsonx.Path(v, "stopFinder", "points", "point")
	o, ok := point.(*jsonx.Ordered)
	if !ok {
		return nil, nil
	}
	return o, nil
}

// Haltestellen searches for stops, mirroring Efa::haltestellen().
func (c *Client) Haltestellen(search string) ([]Option, error) {
	v, err := c.get("XSLT_STOPFINDER_REQUEST", []queryParam{
		{key: "language", value: "de"},
		{key: "outputFormat", value: "JSON"},
		{key: "coordOutputFormat", value: "WGS84[DD.ddddd]"},
		{key: "itdLPxx_usage", value: "origin"},
		{key: "useLocalityMainStop", value: "true"},
		{key: "SpEncId", value: "0"},
		{key: "locationServerActive", value: "1"},
		{key: "stateless", value: "1"},
		{key: "type_sf", value: "any"},
		{key: "anyObjFilter_sf", value: "2"},
		{key: "anyMaxSizeHitList", value: "50"},
		{key: "name_sf", value: search},
	})
	if err != nil {
		return nil, err
	}
	var options []Option
	switch points := jsonx.Path(v, "stopFinder", "points").(type) {
	case []any:
		for _, p := range points {
			if point, ok := p.(*jsonx.Ordered); ok {
				ref := jsonx.Path(point, "ref")
				if refObj, ok := ref.(*jsonx.Ordered); ok {
					gid, _ := refObj.Get("gid")
					name, _ := point.Get("name")
					options = append(options, Option{
						Key:   stringOf(gid),
						Label: stringOf(name),
					})
				}
			}
		}
	case *jsonx.Ordered:
		ref := jsonx.Path(points, "point", "ref")
		if refObj, ok := ref.(*jsonx.Ordered); ok {
			gid, _ := refObj.Get("gid")
			name, _ := jsonx.Path(points, "point").(*jsonx.Ordered).Get("name")
			options = append(options, Option{
				Key:   stringOf(gid),
				Label: stringOf(name),
			})
		}
	}
	return options, nil
}

// RouteArgs mirrors the route command arguments.
type RouteArgs struct {
	Date        string
	Time        string
	Start       string
	Destination string
	Mode        string
}

// Route plans a trip, mirroring Efa::route().
func (c *Client) Route(args RouteArgs) (any, error) {
	return c.get("XSLT_TRIP_REQUEST2", []queryParam{
		{key: "itdDate", value: args.Date},
		{key: "itdTime", value: args.Time},
		{key: "language", value: "de"},
		{key: "sessionID", value: "0"},
		{key: "outputFormat", value: "JSON"},
		{key: "type_origin", value: "stop"},
		{key: "name_origin", value: args.Start},
		{key: "type_destination", value: "stop"},
		{key: "name_destination", value: args.Destination},
		{key: "itdTripDateTimeDepArr", value: depArrCode(args.Mode)},
	})
}

// depArrCode maps the user-facing "Departure"/"Arrival" mode to the short
// code the EFA API expects for itdTripDateTimeDepArr.
func depArrCode(mode string) string {
	if mode == "Arrival" {
		return "arr"
	}
	return "dep"
}

// Abfahrt returns the departures for a stop, mirroring Efa::abfahrt().
// The limit is passed through as a string, exactly like the PHP
// application passes the raw option value into the URL.
func (c *Client) Abfahrt(stop string, limit string, isGID bool) ([]any, error) {
	gid := stop
	if !isGID {
		haltestelle, err := c.Haltestelle(stop)
		if err != nil {
			return nil, err
		}
		if haltestelle != nil {
			ref := jsonx.Path(haltestelle, "ref")
			if refObj, ok := ref.(*jsonx.Ordered); ok {
				g, _ := refObj.Get("gid")
				gid = stringOf(g)
			} else {
				gid = ""
			}
		} else {
			gid = ""
		}
	}
	v, err := c.get("XML_DM_REQUEST", []queryParam{
		{key: "laguage", value: "de"},
		{key: "typeInfo_dm", value: "stopID"},
		{key: "deleteAssignedStops_dm", value: "1"},
		{key: "useRealtime", value: "1"},
		{key: "mode", value: "direct"},
		{key: "outputFormat", value: "rapidJSON"},
		{key: "limit", value: limit},
		{key: "nameInfo_dm", value: gid},
	})
	if err != nil {
		return nil, err
	}
	return toAnySlice(jsonx.Path(v, "stopEvents")), nil
}

func toAnySlice(v any) []any {
	if arr, ok := v.([]any); ok {
		return arr
	}
	return []any{}
}

func stringOf(v any) string {
	if s, ok := v.(string); ok {
		return s
	}
	return ""
}
